geps/gep-1651.md

```diff
@@ -0,0 +1,337 @@
+
+## Introduction
+
+One of the early feature requests for Knative was the ability to deploy an
+application using Knative's HTTP routing support, but make it only available
+within the cluster. I want to be able to specify both the "cluster"
+(service.namespace.svc) and "external" (service.namespace.example.com).
+Gateways using the same GatewayClass on the cluster, but ensure that the
+"cluster" service is only routable within the cluster. This would greatly
+simplify deployment for users over the instructions we have today.
+
+Likewise another use case is to provide load balancing capabilities within a virtual
+private network. Different IaaS providers offer private load balancers to support
+these use cases.
+
+## API
+
+We propose adding a new `routability` field under the `spec.infrastructure` stanza of a Gateway.
+
+### Predefined Routability Values
+
+Implementations MAY implement the following values for 'routability' and MUST abide by
+their defined semantics.
+
+Value | Scope
+-|-
+`Public`|The address is routable on the public internet
+`Private`|The address is routable inside a private network larger than a single cluster (ie. VPC) and MAY include RFC1918 address space
+`Cluster`|The address is routable inside the [cluster's network](https://kubernetes.io/docs/concepts/cluster-administration/networking/#how-to-implement-the-kubernetes-network-model)
+
+Values can be compared semantically - `Public` has a larger scope than `Private`, while `Private` has a larger scope than `Cluster`.
+
+### Vendor prefixed values
+
+Implementations can define custom 'routability' values by specifying a vendor prefix followed
+by a slash `/` and a custom name ie. `com.example.com/my-routability`.
+
+Comparing vendor prefixed scopes with the pre-defined ones in implementation specific.
+
+### Default Routability
+
+The default value of `routability` is implementation specific. It is RECOMMENDED that
+the default `routability` remains consistent for Gateways with the same
+`gatewayClassName`.
+
+Implementations MAY leave the `spec.infrastructure.routability` property unset and instead
+signal the default routability using the Gateway's `status.addresses`. See 'Status Addresses`
+for more details.
+
+### Mutability
+Implementations MAY prevent end-users from updating the `routability` value of a Gateway. If
+updates are allowed the semantics and behaviour will depend on the underlying implementation.
+
+If a Gateway is mutated but does not support the desired routability it MUST set the conditions
+`Accepted`, `Ready`, `Programmed` to `False` with `Reason` set to `UnsupportedRoutability`.
+
+### Go
+
+```go
+
+type GatewayRoutability string
+
+const (
+  GatewayRoutabilityPublic  GatewayRoutability  = "Public"
+  GatewayRoutabilityPrivate GatewayRoutability  = "Private"
+  GatewayRoutabilityCluster GatewayRoutability  = "Cluster"
+)
+
+type GatewaySpec struct {
+  // Infrastructure defines infrastructure level attributes about this Gateway instance.
```

Comment:
General comment on the API Spec in this GEP - there's a lot more detail in the markdown than in the proposed API spec. Now that we're close to finalized on this approach, it would be helpful to update this section to have significantly more detail.

Reply:
Is this necessary here? I figured when I do a follow up PR modifying the types I'd move all the details in the markdown into godoc

Reply:
Done